### PR TITLE
feat(openai): add Responses API transport

### LINKED ISSUE #103: fix(openai): support reasoning models with function tools
## Problem

GAI v0.5.9 cannot use an OpenAI reasoning model with function tools when the model rejects that combination on the Chat Completions API.

A live agent run with:

- provider: OpenAI
- model: `gpt-5.6-terra`
- one native function tool
- streaming enabled

fails after all retries with:

```text
Function tools with reasoning_effort are not supported for gpt-5.6-terra in /v1/chat/completions.
To use function tools, use /v1/responses or set reasoning_effort to 'none'.
```

The HTTP/SSE transport can already be open by the time the provider rejects the request, so applications receive an error inside an otherwise successful HTTP 200 event stream.

## Current behavior

The OpenAI adapter:

1. Uses `Chat.Completions.New` / `NewStreaming` for all OpenAI models.
2. Maps every request's reasoning effort onto `params.ReasoningEffort`.
3. Defines only `low`, `medium`, and `high`; applications cannot explicitly request `none`.
4. Advertises native tools for the model, so the agent sends its function definitions without detecting this model/transport/reasoning incompatibility locally.

Relevant code:

- `ai/openai/model.go`: Chat Completions transport and reasoning-effort mapping.
- `ai/capability.go`: provider-neutral reasoning-effort values.
- `agent/agent.go` and `loop/loop.go`: tool definitions and model requests.

## Expected behavior

GAI should either support this valid OpenAI model/tool combination or reject it locally with a typed capability error before making the provider request.

## Proposed direction

Support one or both provider-compatible paths:

1. Add an explicit provider-neutral `ReasoningEffortNone` value and map it to OpenAI's `reasoning_effort: "none"`.
2. Add an OpenAI Responses API transport that preserves streaming text, function calls/results, native message history, usage, and provider errors.

Transport and reasoning selection should be explicit and capability-aware rather than inferred solely from a model-name family. This should compose with the capability work in #83.

## Acceptance criteria

- [ ] An OpenAI reasoning model can stream a response while function tools are present using a supported transport/configuration.
- [ ] Applications can explicitly select `reasoning_effort: none` when the model supports it.
- [ ] If the selected model, transport, tools, and reasoning mode are incompatible, GAI returns a typed local capability error before the provider request.
- [ ] Empty/default reasoning configuration does not accidentally serialize an unsupported reasoning mode.
- [ ] Tests cover streaming with a function tool and explicit `none`.
- [ ] If Responses API support is added, tests cover tool-call/tool-result continuation and terminal provider errors.
- [ ] Provider documentation explains which OpenAI transport and reasoning modes are used.

### PR DESCRIPTION
**Pull request by SecondBraincell**
Closes #103

## Summary
- add an explicit OpenAI Responses transport with native history, function continuation, streaming, usage, and terminal error mapping
- support explicit reasoning_effort none and preflight GPT-5 Chat Completions tool/reasoning incompatibility
- document transport selection and add regression coverage

## Validation
- go test ./...
- go vet ./...

## Summary by CodeRabbit

* **New Features**
  * Added optional OpenAI Responses API support for synchronous and streaming generation.
  * Added support for Responses-specific tools, reasoning settings, response formats, usage details, and tool-call results.
  * Expanded reasoning-effort options, including none, minimal, extra-high, and maximum.
  * Added configurable transport selection, with Chat Completions remaining the default.
  * Added GPT-5 model compatibility guidance.

* **Bug Fixes**
  * Added validation for unsupported reasoning and tool combinations.
  * Improved handling of streaming errors, cancellation, invalid tool data, and provider failures.

* **Documentation**
  * Expanded OpenAI setup and model capability guidance.

<h3>Greptile Summary</h3>

This PR adds an explicit OpenAI Responses API transport alongside the existing Chat Completions transport, gated via `WithResponsesTransport()`. It wires up native tool-call/result history, reasoning items in `ThoughtSignature`, streaming, usage mapping, and terminal-error handling, plus extends `ReasoningEffort` with `none`, `minimal`, `xhigh`, and `max` values and adds a preflight guard that blocks GPT-5 + tools + explicit reasoning effort on the Chat Completions endpoint.

- **New `responses.go`**: implements `generateResponses` / `generateResponsesStream` / `buildResponsesParams` / `mapResponsesMessages` with full tool-continuation, reasoning-item round-trip, and error fallback paths.
- **`model.go` updates**: adds `validateTransport` guard, `isGPT5ReasoningFamily`, `gpt5ReasoningEfforts` for per-revision effort lists, and updates `isOpenAIReasoningFamily` to delegate to the new helper; `buildChatCompletionParams` now accepts all seven effort values.
- **`provider.go` / `capability.go`**: `New` becomes variadic for `Option` functions; five new `ReasoningEffort` constants are exported.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the new Responses transport is well-isolated behind an opt-in option, the Chat Completions default path is unchanged, and the prior unguarded streaming-ThoughtSignature gap is now correctly addressed.

All six changed files have comprehensive regression tests covering tool continuation, reasoning-item round-trips, streaming ThoughtSignature propagation, terminal error fallback, refusal text, and the Chat Completions preflight guard for both none and non-none reasoning effort. The transport dispatch is additive and the default path is unmodified.

**Files Needing Attention:** No files require special attention.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| ai/openai/responses.go | New file implementing the Responses API transport; sync and streaming paths both correctly accumulate reasoning items into ThoughtSignature before emitting tool-call tokens, addressing the key correctness concern for multi-turn continuation. |
| ai/openai/model.go | Adds validateTransport guard (blocks GPT-5 + tools + reasoning on Chat Completions), per-revision gpt5ReasoningEfforts dispatch, and the transport-dispatch branches in Generate/GenerateStream; logic is correct and well-tested. |
| ai/openai/provider.go | Adds Transport type, WithResponsesTransport option, and makes New variadic; nil-option guard is a safe touch. |
| ai/capability.go | Exports five new ReasoningEffort constants; additive change with no breaking impact on existing callers. |
| ai/openai/model_test.go | Extensive new coverage for Responses transport: tool continuation, reasoning-item round-trip, streaming ThoughtSignature, refusal text, error fallback, store=false, and include=reasoning.encrypted_content; all edge cases from prior review threads now have regression tests. |
| ai/openai/provider_test.go | Updates existing dynamic-descriptor tests to use gpt-5.6-terra (a known revision with 6 effort levels) so the assertions match the new gpt5ReasoningEfforts logic; no logic change. |

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Generate / GenerateStream"] --> B["ai.ValidateModelRequest"]
    B --> C["validateTransport"]
    C -->|"GPT-5 + tools + reasoning effort\non Chat Completions"| D["ErrUnsupportedCapability"]
    C -->|"ok"| E{"provider.transport?"}
    E -->|"TransportResponses"| F["generateResponses /\ngenerateResponsesStream"]
    E -->|"TransportChatCompletions (default)"| G["buildChatCompletionParams"]
    F --> H["buildResponsesParams"]
    H --> I["mapResponsesMessages\n(native tool/reasoning history)"]
    I --> J["Responses.New /\nResponses.NewStreaming"]
    J -->|"sync"| K["responseFromResponses\n(collect reasoning items → ThoughtSignature)"]
    J -->|"stream"| L["event loop\nreasoning accumulator\n→ ThoughtSignature per call"]
    G --> M["mapNativeMessages"]
    M --> N["ChatCompletion.New /\nChatCompletion.NewStreaming"]
```
</details>

<sub>Reviews (6): Last reviewed commit: ["Merge branch main into fix/openai-respon..."](https://github.com/lace-ai/gai/commit/794a4cb2d8960fb0b3c939fd04850c53719303d8) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=49145411)</sub>

**Context used:**

- Context used - AGENTS.md ([source](https://app.greptile.com/laceai/github/lace-ai/gai/-/custom-context?memory=173cdb2a-e8ea-489c-91ae-aa613f5b79be))

### CODE REVIEW

**Comment by Samuelk0nrad** [comment]
@codex review

File: ai/capability.go [hunk 1]
```diff
@@ -217,6 +217,7 @@ type ResponseFormat struct {
 type ReasoningEffort string
 
 const (
+	ReasoningEffortNone   ReasoningEffort = "none"
 	ReasoningEffortLow    ReasoningEffort = "low"
 	ReasoningEffortMedium ReasoningEffort = "medium"
 	ReasoningEffortHigh   ReasoningEffort = "high"
```
Comment (Samuelk0nrad) [inline ai/capability.go]: there are some models that support even higher reasoning efforts like xhigh, max, ultra and so on right?

**Review by Samuelk0nrad — CHANGES_REQUESTED** [review]

**Comment by Samuelk0nrad** [comment]
@codex review

File: ai/openai/model.go [hunk 2]
```diff
@@ -251,11 +269,25 @@ func openAIDescriptor(model string) ai.ModelDescriptor {
...
+}
+
+func isOpenAIReasoningFamily(model string) bool {
+	if isGPT5ReasoningFamily(model) {
 		return true
 	}
+	name := strings.ToLower(strings.TrimSpace(model))
 	if len(name) < 2 || name[0] != 'o' {
 		return false
 	}
```
Comment (Samuelk0nrad) [inline ai/openai/model.go:309]: Reasoning efforts are advertised too broadly

The descriptor currently gives every model beginning with gpt-5 all seven values


* GPT-5: `minimal`, `low`, `medium`, `high`
* GPT-5.1: `none`, `low`, `medium`, `high`
* GPT-5.6: `none`, `low`, `medium`, `high`, `xhigh`, `max`

[OpenAI Developers](https://developers.openai.com/api/docs/models/gpt-5?utm_source=chatgpt.com)

**Comment by Samuelk0nrad** [comment]
@SecondBraincell fix the merge conflict (merge main into this branch) do not add to many detailed to the readme

**Comment by SecondBraincell** [comment]
Merged `main` into this branch and resolved the README conflict by retaining `main`’s version without adding extra Responses API detail.

`go test ./...` passes. Head is `794a4cb2d8960fb0b3c939fd04850c53719303d8`; GitHub CI is queued.

**Review by Samuelk0nrad — APPROVED** [review]
